### ISSUE TITLE
Inject services in custom filter

### ISSUE THREAD

**Issue by ssougnez** [dotliquid/dotliquid#402]
I'd like to create a custom translate filter. For this, I need to be able to inject services in the filter, but as they are static method, I can't find a way to do so.

Is there some way to do this? Maybe it would be interesting to have `FilterFactory` or something like that.

**Comment by microalps** [comment]
Filters can have Context context as one of the parameters. With this you can pass any arbitrary data, even without specifying it is 'safe' or inheriting a drop.

**Comment by microalps** [comment]
We are closing this issue as the question is answered. Please reopen of you have  further questions or feeback.

**Comment by pmiddleton** [comment]
@microalps - I am trying to do this same thing.  I need to pass a service into a filter.  I am trying your suggestion but am having some issues.

I created a new context and am trying to pass my service in via the indexer on the context.  Inside my filter I am only able to access the service via the context if the service type is registered as a safetype.  When I get the service out it is wrapped in a DropProxy.  

I see no other obvious way to pass the service via the context nor how to unwrap the service from the DropProxy.  Can you give an example of how to do this?

**Comment by microalps** [comment]
Here is some sample code written in LinqPad. SiteConfig is a class that is not a drop or otherwise whitelisted.
```c#
void Main()
{
	Template.RegisterFilter(typeof(FilterExtensions));
	Template.NamingConvention = new DotLiquid.NamingConventions.CSharpNamingConvention();
	var template = Template.Parse("{{ Date | ShortFormat }}");
	var renderParameters = new RenderParameters(CultureInfo.CurrentCulture) { 
		LocalVariables = Hash.FromAnonymousObject(new { Date = DateTime.Now }),
		Registers = Hash.FromAnonymousObject(new { Config = new SiteConfig() }) 
	};
	template.Render(renderParameters).Dump();
}

public class SiteConfig
{
	public string ShortFormat { get; set; } = "MM/yyyy";
	public string LongFormat { get; set; } = "G";
}

public class FilterExtensions
{
	public static string ShortFormat(Context context, DateTime input)
	{
		return input.ToString(((SiteConfig)context.Registers["Config"]).ShortFormat);
	}
}
```

**Comment by pmiddleton** [comment]
@microalps Thanks.

I tried that approach originally but ran into an issue.  When you do things that way then the Context of renderParameters is null so you don't get back any errors from rendering.  If you assign a context to renderParameters then the LocalVariables are ignored. 

In order to get errors I needed to create a new context and add my services to the Environment property.  Then I register the filter like this

```csharp
 Template.RegisterSafeType(typeof(FilterExtensions), new[] { "*" }, o => o);
```

the ```o => o ``` unwrapped the filter from the DropProxy.

I don't see a way to get the render errors out using your approach.
